### PR TITLE
Implement no-loop rule attribute

### PR DESCRIPTION
**Pull request by markbjerke**
These changes allow for the no-loop rule attribute.  This is based upon the attribute of the same name in Drools.

The work involves enhancement request, #141 .
@mention doug-martin

### CODE REVIEW

File: .gitattributes [hunk 1]
```diff
@@ -0,0 +1,17 @@
+# Auto detect text files and perform LF normalization
```
Comment (doug-martin) [inline .gitattributes]: Can you please remove this file

File: lib/agenda.js [hunk 2]
```diff
@@ -162,6 +162,14 @@ module.exports = declare(EventEmitter, {
 
         insert: function (node, insert) {
             var rule = this.rules[node.name], nodeRule = node.rule, agendaGroup = nodeRule.agendaGroup;
+            if (nodeRule.noLoop) {
+                if (rule.noLoop[insert.hashCode]) {
+                    return;
+                }
```
Comment (doug-martin) [inline lib/agenda.js]: Formatting 
```
} else {
```

File: lib/parser/nools/tokens.js [hunk 3]
```diff
@@ -94,6 +94,20 @@ var ruleTokens = {
         };
     })(),
 
+    noLoop: (function () {
+        var noLoopRegexp = /^(noLoop|no-loop)\s*:\s*(-?true|false)\s*[,;]?/;
```
Comment (doug-martin) [inline lib/parser/nools/tokens.js]: what is the `-?` for in this case when would you have a `-true`

File: readme.md [hunk 4]
```diff
@@ -1003,6 +1004,33 @@ flow1
...
+
+```javascript
+rule Hello {
+    no-loop: true;
+    when {
+        m: Message m.name like /hello/;
+    }
+    then { modify(m, function() {
+		m.text = 'hello world'
+	});
```
Comment (doug-martin) [inline readme.md]: I think your missing a `}` here

```
then {
    modify(m, function () {
        m.text = 'hello world';
    });
}
```

File: test/flow/noLoop.test.js [hunk 5]
```diff
@@ -0,0 +1,67 @@
...
+                    this.modify(m);
+                }
+            });
+        });
+
+    var noolsSource =  "rule 'Hello3' { no-loop: true; when {m: Message m.name =~/Hello/;}then {modify(m, function () { this.name = 'Hello World'; });}}";
+
+    var flow3 = nools.compile(noolsSource, {
+        name: 'testDsl'
+        ,define: {
```
Comment (doug-martin) [inline test/flow/noLoop.test.js]: comma should be on the previous line

File: test/flow/noLoop.test.js [hunk 6]
```diff
@@ -0,0 +1,67 @@
...
+    var flow1 = nools.flow("noLoop1", function () {
+        
+        this.rule("Hello2", { noLoop: true }, [Message, "m", "m.name =~ /Hello/"], function (facts) {
+            var m = facts.m;
+            m.name = 'Hello World';
+            this.modify(m);
+        });
+    }),
+	
+        flow2 = nools.flow("noLoop2", function () {
```
Comment (doug-martin) [inline test/flow/noLoop.test.js]: This should be in line with the flow1 definition

For example

```javascript
var flow1 = ...,
    flow2 = ...;
```

File: test/flow/noLoop.test.js [hunk 7]
```diff
@@ -0,0 +1,67 @@
...
+            fired1.push(name);
+        }),
+        session2 = flow2.getSession(new Message("Hello")).on("fire", function (name) {
+            fired2.push(name);
+        }),
+        session3 = flow3.getSession(new Message("Hello")).on("fire", function (name) {
+            fired3.push(name);
+        });
+        return session1.match()
+            .then(function () {
```
Comment (doug-martin) [inline test/flow/noLoop.test.js]: Please reformat this block, very hard to figure out how the braces match up.

**Comment by markbjerke** [comment]
I'm closing these in order to factor out changes in the code that support feature from the feature itself.

**Pull request closed without merging**